brickschema/validate.py

```diff
@@ -0,0 +1,423 @@
+"""
+The `validate` module implements a wrapper of `pySHACL`_ to
+validate an ontology graph against the default Brick Schema constraints (called *shapes*) and user-defined shapes.
+
+.. _`pySHACL`: https://github.com/RDFLib/pySHACL
+"""
+import sys
+import argparse
+import logging
+from rdflib import Graph, Namespace, URIRef, BNode, Literal
+from rdflib.plugins.sparql import prepareQuery
+from .namespaces import BRICK, A, RDF, RDFS, BRICK, BSH, SH, SKOS, bind_prefixes
+import pyshacl
+import io
+import pkgutil
+
+class Validate():
+    """
+    Validates a data graph against Brick Schema and basic SHACL constraints for Brick.  Allows extra
+    constraints specific to the user's ontology.
+    """
+
+    # build accumulative namespace index from participating files
+    # build list of violations, each is a graph
+    def __init__(self, useBrickSchema=True, useDefaultShapes=True):
+        self.log = logging.getLogger()
+        self.log.setLevel(logging.DEBUG if hasattr(sys, '_called_from_test') else logging.WARNING)
+        self.log.info('Validate init')
+
+        self.namespaceDict = {}
+        self.brickG = Graph()
+        self.shapeG = Graph()
+
+        if useBrickSchema:
+            data = pkgutil.get_data(__name__, "ontologies/Brick.ttl").decode()
+            self.brickG.parse(source=io.StringIO(data), format='turtle')
+            self.__buildNamespaceDict(self.brickG)
+
+            # Remove rdfs:domain and rdfs:range.  The modified
+            # ontology will be used for pySHACL reasoning.
+            # See DESIGN.md for more discussion.
+
+            self.brickG.update('DELETE { ?s rdfs:domain ?o .} WHERE { ?s rdfs:domain ?o . }',
+                               initNs=self.namespaceDict)
+            self.brickG.update('DELETE { ?s rdfs:range ?o .} WHERE { ?s rdfs:range ?o . }',
+                               initNs=self.namespaceDict)
+
+        if useDefaultShapes:
+            data = pkgutil.get_data(__name__, "ontologies/BrickShape.ttl").decode()
+            self.shapeG.parse(source=io.StringIO(data), format='turtle')
+            self.__buildNamespaceDict(self.shapeG)
+
+
+    def validate(self, data_graph, shacl_graph=None, ont_graph=None,
+                 inference='rdfs', abort_on_error=False, advanced=True,
+                 meta_shacl=True, debug=False):
```

Comment:
See my comment on the `README` above: `validate` should return a Result object that contains the `conforms`, `graph` and `result text` properties. This Result class does not need to be super complex: it just provides a logical grouping of the outputs

Reply:
See my comments before why it's done this way.

Reply:
It's jarring and confusing to have both the stateless and stateful APIs combined in this way with no clear reason why one would want to provide a single shacl/ontology definition graph at the time of calling. I don't think preserving the pyshacl API is sufficient grounds, because it's not a widely known interface